Honeycomb lattice & LatticePhysics support (WIP)

ffreyer opened the pull request:
Adds a Honeycomb lattice with nearest and next nearest neighbors. The checkerboard decomposition doesn't include NNN yet.

crstnbr commented:
On a side note, it would be cool to have an interface to Jan's [LatticePhysics.jl](https://github.com/janattig/LatticePhysics.jl) at some point, so we get all his lattices for free.

ffreyer commented:
Added next nearest neighbors to the checkerboard setup. Could you check if I did it correctly?

ffreyer commented:
The Hubbard model should now work with LatticePhysics lattices. I made the following changes:

* added a function `numSites(::MonteCarlo.Lattice)` (equivalent to `LatPhysBase.numSites`) and used it instead of `l.sites`
* changed HubbardModel to accept `LatPhysBase.AbstractLattice`, changed `hopping_matrix` to dispatch on the Lattice type via new `_hopping_matrix(lattice, mu, t)`
* added `build_checkerboard(::LatPhysBase.AbstractLattice)`. For a 4x4 square lattice, this gives the same result when reordering bonds, i.e. with `bonds = [l.bonds[4i+j] for i in 0:15 for j in [3, 1]]` in `build_checkerboard()`

carstenbauer commented on src/abstract.jl:
```diff
@@ -29,6 +29,9 @@ However, any concrete Lattice type should have at least the following fields:
 """
 abstract type Lattice end
 
+numSites(l::Lattice) = l.sites
```
The camel casing is against the Julia style guide rules (and I personally don't like it much). If we need this getter method, we should better call it `num_sites` (or maybe even `length`?) and define `num_sites(lattice::LatPhysBase.AbstractLattice) = numSites(lattice)`.

carstenbauer replied:
Another idea that comes to mind is adjusting our (abstract) definition of a lattice to make it compatible with the corresponding one in LatticePhysics. After all, we don't yet have many lattices here.

carstenbauer replied:
Also, just to be safe, we should prefix with `@inline`.

ffreyer replied:
Names have been changed, `@inline` has been added as of d9e65ad

carstenbauer commented on src/flavors/DQMC/abstract.jl:
```diff
@@ -5,53 +5,109 @@ struct CheckerboardTrue <: Checkerboard end
+                    belong to which checkerboard group.
+    * `n_groups::Int`: number of checkerboard groups.
+"""
+function build_checkerboard(l::LatPhysBase.AbstractLattice)
```
We should try to avoid copying code here. Both `build_checkerboard` methods are almost identical. We can probably define more methods like `num_bonds` etc. similar to `num_sites` above and implement one single generic version.

ffreyer replied:
Back to one `build_checkerboard()` as of 7387b0c

carstenbauer commented on src/flavors/DQMC/slice_matrices.jl:
```diff
@@ -53,7 +53,7 @@ const DQMC_CBTrue = DQMC{M, CheckerboardTrue} where M
 function slice_matrix(mc::DQMC_CBTrue, m::Model, slice::Int,
 					power::Float64=1.)
-  M = eye(heltype(mc), m.flv*m.l.sites)
+  M = eye(heltype(mc), m.flv*numSites(m.l))
```
Is the `eye` function here even functional? It has been dropped from `Base` in 0.7 and we don't seem to define it anywhere. Could you replace those instances by something based on `LinearAlgebra.I`?

carstenbauer commented on src/lattices/Honeycomb.jl:
```diff
@@ -0,0 +1,112 @@
+"""
+Two dimensional square lattice.
+"""
+mutable struct HoneycombLattice <: AbstractCubicLattice
```
Why do we need an explicit `HoneycombLattice` if we support LatticePhysics?

ffreyer replied:
I'll need it in the future, so I thought I might as well push it. I can remove it from the pr if you want...

carstenbauer commented on src/models/HubbardAttractive/HubbardModelAttractive.jl:
```diff
@@ -15,7 +15,10 @@ with linear system size `L`. Additional allowed `kwargs` are:
 """
-@with_kw_noshow mutable struct HubbardModelAttractive{C<:AbstractCubicLattice} <: Model
+@with_kw_noshow mutable struct HubbardModelAttractive{
+        LT <: Union{AbstractCubicLattice, LatPhysBase.AbstractLattice}
```
I have to think about whether there is a smarter way than to use `Union` everywhere. Until then, it might make sense to define something like `const SomeLattice = Union{AbstractCubicLattice, LatPhysBase.AbstractLattice` once.

carstenbauer replied:
More importantly, why allow a `LatPhysBase.AbstractLattice` here but only a `AbstractCubicLattice` in `hopping_matrix` below? Note that the former would work in `hopping_matrix` since we use `l.neighs` to access the neighbors their - a field which Jan's lattices don't have. We need a general solution here if we want to support all LatticePhysics lattices. Does Jan have a `getNeighbors` method?

ffreyer replied:
There are two definitions, one for `LatPhysBase.AbstractLattice` and one for `AbstractCubicLattice`. I'm using `bonds(::AbstractLattice)` instead of `l.neighs` there.

ffreyer replied:
Back to one hopping matrix as of 7387b0c, Union has been changed to `AbstractLattice` as of 3cbe4fa.

crstnbr commented:
One more thing, it'd be good to add some tests for lattices.

ffreyer commented:
* `Lattice` is now `AbstractLattice`
* `numSites` is now `length`
* `neighbors(lattice, drop_repeats::Val{true/false}` added
* ^ thoses things are now in `lattices/abstract.jl`
* build_checkerboard now works the same for LatPhys lattices and MonteCarlo Lattices

carstenbauer commented on src/lattices/abstract.jl:
```diff
@@ -0,0 +1,62 @@
+        for b in LatPhysBase.bonds(l)
+        if LatPhysBase.from(b) < LatPhysBase.to(b)
+    )
+end
```
As a general note, keep in mind that we certainly want to be able to initially create a static lookup table of neighbors in MC to get O(1) lookup.

ffreyer replied:
I don't think LatPhys has something like that. So we should add a thin wrapper which generates the `neighs` from MonteCarlo.jl, containing the same information as `neighbors(lattice, Val(false))`. I don't think we need one for `neighbors(lattice, Val(true))` though.

Imo it would be useful to have
* `all_neighbors(lattice, site)` -> Vector of targets
* `no_repeated_neighbors(lattice)` -> Iterable of (source, target) pairs

I don't think there is a reason to keep an `all_neighbors(lattice)` around.

ffreyer replied:
I added a wrapper for LatPhys lattices and a method `neighbors(lattice, site)`

carstenbauer replied:
One more thing: do we explicitly want to depend on LatPhys*.jl (in Project.toml)? Maybe it's better to use Requires.jl to only load the LatticePhysics stuff when, say, LatPhysBase.jl has been explicitly loaded.

ffreyer replied:
I think that's a good idea. I'll make another commit for this.

Btw, do you want me to remove the Honeycomb lattice from this?

carstenbauer replied:
I guess it doesn't hurt to leave it in there for now. We can remove it later if you want/have to. BTW: I'm currently revisiting MonteCarlo.jl on a general (somewhat higher) level. So don't wonder about some changes leading to merge conflicts. We'll fix them later.

ffreyer commented:
TODO

- [ ] Tests for `Honeycomb()`
- [ ] Tests for `ALPSLattice()`. I guess we should take some small lattice, build it by hand and compare it to the parsed version?
- [ ] update documentation. Should we do this here or in a different pr?

crstnbr commented:
I'm rebasing this (which is a big mess :D). Please don't push anything here in the meantime.

crstnbr commented:
At least, tests are working locally now.

* [ ] Honeycomb should be immutable
* [ ] resolve `nsites` vs `length` mess
* [ ] overthink `neighbors` approach

crstnbr commented:
> * Tests for `ALPSLattice()`. I guess we should take some small lattice, build it by hand and compare it to the parsed version?

We can use https://github.com/crstnbr/dqmc/blob/master/lattices/gen_square_lattice_xml.jl

carstenbauer commented on src/lattices/abstract.jl:
```diff
@@ -0,0 +1,66 @@
+    neighbors(l::AbstractLattice, drop_repeats::Val{false})
+
+Returns an iterator over bonds, given as tuples (source index, target index). If
+`drop_repeats = Val(false)` bonds are assumed to be directed, i.e. both
```
`directed` seems much better than `drop_repeats` to me.

ffreyer replied:
I agree

carstenbauer commented on src/lattices/abstract.jl:
```diff
@@ -0,0 +1,66 @@
+"""
+@inline function neighbors(l::AbstractLattice, drop_repeats::Val{false})
+    (
+        [source, target] for source in 1:length(l)
```
Why return a vector (not a tuple)?

ffreyer replied:
Because the other method returns vectors. I thought it'd be better to be consistent. (Yet two methods return iterators of vectors and one a vector integers...)

carstenbauer commented on src/lattices/abstract.jl:
```diff
@@ -0,0 +1,66 @@
+
+# Typically, you also want to implement
+
+#     - `neighbors_lookup_table(lattice)`: return a neighbors matrix where
```
More important than getting the neighbors as iterators is getting them as a lookup table (since we want to minimize neighbor lookup costs during Monte Carlo).

ffreyer replied:
I did some benchmarks. (with a dummy lattice, should be equivalent)

* `for trg in lattice.neighs[:, src]` is just as fast as `for trg in neighbors(lattice, src)`
* `for (src, trg) in neighbors(lattice, Val(false))` is 4x slower than `for src in 1:N; for trg in neighbors(lattice, src)`, but I think we only use this method to generate the checkerboard.
* `neighs[1] + neighs[2] + ...` with `neighs = neighbors(lattice, src)` is about 2.5x slower than indexing `lattice.neighs` directly, but
* `neighbors(lattice, src, 1) + neighbors(lattice, src, 2) + ...` is again just as fast, with `@inline neighbors(lattice, src, i) = lattice.neighs[i, src]`

For the Ising model all the general methods use either `neighbors(lattice, site_index)` (local, global updates) or `neighbors(lattice, Val(true)` (energy). The former performs fine, the latter doesn't use a lookup table.
Only the methods specialized on the `SquareLattice` suffer some performance penalties here, since there is no `neighbors(lattice, src, i)`. But I would argue that it's fine for those methods to just access `lattice.neighs` directly, because they only work for one specific lattice we implement.

As you mentioned below, `hopping_matrix` uses `neighbors(lattice, Val(false))`. I'm not too bothered by this, as it's only called once. But we can easily change this to use `neighbors(lattice, site_index)` to get better performance.

carstenbauer replied:
```julia
using BenchmarkTools, MonteCarlo
import MonteCarlo: neighbors, HoneycombLattice, AbstractLattice

l = HoneycombLattice(20)

eachsite(l::AbstractLattice) = Base.OneTo(length(l))

@inline neighbors(lattice, src, i) = lattice.neighs[i, src]

function f_neighbors_site(l)
    x = 0.0
    @inbounds for src in eachsite(l)
        for trg in neighbors(l,src) # allocates
            x += trg
        end
    end
    x
end

function f_neighbors_site_i(l)
    x = 0.0
    @inbounds for src in eachsite(l)
        for i in 1:size(l.neighs, 1)
            trg = neighbors(l, src, i)
            x += trg
        end
    end
    x
end

function f_table_lookup(l)
    x = 0.0
    @inbounds for src in eachsite(l)
        for i in 1:size(l.neighs, 1)
            trg = l.neighs[i,src]
            x += trg
        end
    end
    x
end

######## Microbenchmarks
@btime neighbors(l,1);
@btime $(l.neighs)[:,1];

# julia> @btime neighbors(l,1);
#   57.580 ns (1 allocation: 112 bytes)

# julia> @btime $(l.neighs)[:,1];
#   40.482 ns (1 allocation: 112 bytes)

@btime neighbors(l,1)[3];
@btime $(l.neighs)[3,1];

# julia> @btime neighbors(l,1)[3];
#   80.249 ns (2 allocations: 128 bytes)

# julia> @btime $(l.neighs)[3,1];
#   1.799 ns (0 allocations: 0 bytes)


######## Context benchmarks
@btime f_neighbors_site($l);
@btime f_table_lookup($l);
@btime f_neighbors_site_i($l); # as fast as f_table_lookup because it is literally identical after inlining

# julia> @btime f_neighbors_site($l);
#   63.899 μs (1600 allocations: 175.00 KiB)

# julia> @btime f_table_lookup($l);
#   6.860 μs (0 allocations: 0 bytes)

# julia> @btime f_neighbors_site_i($l);
  # 6.880 μs (0 allocations: 0 bytes)
```

carstenbauer replied:
We define an `AbstractLattice` interface. The question is what mandatory and optional methods do we want as the interface specification.

Proposal:

**Mandatory**

Every lattice should implement those, otherwise no one could use it on the model side for anything.
* `length`: number of lattice sites
* `ndims`: number of dimensions
* `neighbors(lattice[, directed=Val(false)`: Iterator over `(src,trg)` tuples (or `src=>trg` pairs?). If `directed=Val(true)` both `(src,trg)` and `(trg,src)` will be generated.
(This function might be called `bonds(lattice)` in other lattice libraries)

**Optional**

* `neighbors(lattice, site)`: Iterator over neighbors of a given site.
* `neighbor(lattice, site, i)`: Return the site index of the `i`th neighbor of a site
* `neighbors_lookup_table(lattice)`: generate a table for O(1) nearest neighbor lookup
* perhaps `nneighbors(lattice, site)`: number of neighbors of a given site
* perhaps `name`: name of the lattice

We could (and should) provide generic fallback implementation of the optional methods, (based on the mandatory `neighbors(lattice)` iterator).

carstenbauer replied:
Currently, the field `neighs` of all our lattices is a lookup table. Accessing it is as fast as it gets (personally I'd always use a lookup table for Monte Carlo because I cannot be sure that the accessor methods of the lattice are efficient). Generally, we don't want anyone to access fields directly. First of all, the lattice implementation might change at some point but more importantly we want to support lattices from other libraries which might not have an internal nearest neighbors table.

Having a generic `neighbors_lookup_table(lattice)`, even if inefficient, would be a great way to support almost any lattice in `O(1)` speed (at least for nearest neighbors) independent of its implementation. But I agree that `neighbor(lattice, site, i)` should be just as fast in most cases.

ffreyer replied:
My dummy lattice is

```julia
struct Lattice
    N::Int64
    neighs::Matrix{Int64}
end

@inline function neighbors(l::Lattice, site_index::Integer)
    l.neighs[:, site_index]
end

N = 10
neighs = rand(Int64, 4, N)
l = Lattice(N, neighs)
```

With it I get:

```julia
julia> @btime neighbors(l,1);
  26.339 ns (1 allocation: 112 bytes)

julia> @btime $(l.neighs)[:,1];
  25.269 ns (1 allocation: 112 bytes)
```

I get `34ns` and `25-26ns` respectively with your benchmarks though. Defining `neighbors` for the concrete lattice gets it down to `26ns` as well (just slightly slower than accessing `l.neighs`).

ffreyer replied:
```julia
function f_table_lookup(l)
    x = 0.0
    @inbounds for src in eachsite(l)
        for i in 1:size(l.neighs, 1)
            trg = l.neighs[i,src]
            x += trg
        end
    end
    x
end
```

My version uses `for trg in l.neighs[:, src]`, which causes allocations and slows it down to be on par with `for trg in neighbors(l, src)`. With a concrete `HoneycombLattice` in `neighbors()` I'm getting `39.67 μs`, `5.03 μs`, `4.92 μs` for the last three tests, so still a significant slowdown. (Same number of allocations)

carstenbauer replied:
Your `neighbors(l,1)` and `l.neighs[:,1]` allocate. This isn't necessary. We shouldn't have to allocate anywhere in a model to access/iterate over neighbors (although it's fine/acceptable in setting up the Monte Carlo like calculating the hopping matrix).

We should always use `model.neighs[i,1]` or `neighbor(l, 1, i)` (We should change this in IsingModel.jl). The former, having the lookup table on the model side, has the advantage that access is guaranteed to be fast indepentend of the lattice's implementation of `neighbor(l,1,i)`. (`l.neighs` should never be accessed in a model anywhere.)

ffreyer replied:
> * `neighbors_lookup_table(lattice)`: generate a table for O(1) nearest neighbor lookup

I assume you want to do something like `model.neighs = neighbors_lookup_table(lattice)`. I'd rather keep the lattice and the model more seperate and interface through `neighbor(lattice, site, i)`. 
That would require

> * perhaps `nneighbors(lattice, site)`: number of neighbors of a given site

but we should find a different name, one not so easily confused with `neighbors()`.

> * perhaps `name`: name of the lattice

I'm not sure why - the name of the lattice is the type of the lattice already.

carstenbauer replied:
> I assume you want to do something like `model.neighs = neighbors_lookup_table(lattice)`. I'd rather keep the lattice and the model more seperate and interface through `neighbor(lattice, site, i)`.

I want both! If I implement a model I can choose whether I want to use a lookup table (which is guaranteed to be fast) or whether I want to rely on accessor functions of the lattice, like `neighbor(lattice, site, i)`. My personal preference is the former but I'd leave it up to the taste of the person implementing the model.

The reason why I do not want to only have the latter option is that we want to support lattices from external sources like Lattices.jl or LatticePhysics.jl and we might not have full control over the efficienty of accessor methods. The lookup table on the other hand, even if it is costly to generate at first, is always fast.

What we choose for the built-in models, doesn't matter that much. Perhaps we should demonstrate both approaches?

ffreyer replied:
> We should always use `model.neighs[i,1]` or `neighbor(l, 1, i)` (We should change this in IsingModel.jl). The former, having the lookup table on the model side, has the advantage that access is guaranteed to be fast indepentend of the lattice's implementation of `neighbor(l,1,i)`.

We should just pick one direction and go with it. We can change it later if needed.

> I want both!

OK

ffreyer replied:
Btw, what would you need `ndims` for?

carstenbauer replied:
I guess we don't need it. Same for `name`.

carstenbauer replied:
BTW, I dropped the `AbstractCubicLattice` locally. This should be a trait rather than an abstract type. Also, since we don't want to implement `neighbors(lattice, directed)` for all built-in lattices independently, I'm implementing traits for that as well. Something like this:

```julia
# Abstract trait
abstract type LatticeProperties end

"""
    HasNeighborsTable trait

Indicates that a lattice `l` has a neighbors lookup table as a field `l.neighs`.

A neighbors lookup table is a `Matrix{Int}` of size `(nneighbors, nsites)` such that
`l.neighs[i,j]` gives the site index of the `i`th neighbor of site `j`.

A lattice should implement `has_neighbors_table(::MyLattice) = HasNeighborsTable()` to indicate the trait.
"""
struct HasNeighborsTable <: LatticeProperties end

"""
    HasBondsTable trait

Indicates that a lattice `l` has a bonds table as a field `l.bonds`.

A bonds table is a `Matrix{Int}` of size `(nbonds, 3)` where the columns mean
`src`, `trg`, `type`, respectively.

A lattice should implement `has_bonds_table(::MyLattice) = HasBondsTable()` to indicate the trait.
"""
struct HasBondsTable <: LatticeProperties end

has_neighbors_table(::AbstractLattice) = nothing
has_bonds_table(::AbstractLattice) = nothing

@inline neighbors(l::AbstractLattice, directed::Val{true}) = _neighbors(has_neighbors_table(l), l, directed)
@inline neighbors(l::AbstractLattice, directed::Val{false}) = _neighbors(has_bonds_table(l), l, directed)

@inline _neighbors(::Nothing, l::LT, directed::V) where {LT<:AbstractLattice, V<:Val} =
    error("Lattice $(LT) doesn't implement `neighbors(::$(LT), ::$V`")

@inline function _neighbors(::HasNeighborsTable, l::AbstractLattice, directed::Val{true})
    (
        (src, trg) for src in 1:length(l)
            for trg in l.neighs[:, src]
    )
end
@inline function _neighbors(::HasBondsTable, l::AbstractLattice, directed::Val{false})
    (
        l.bonds[i, 1:2] for i in 1:size(l.bonds, 1)
    )
end
```

ffreyer replied:
I wonder if this is more work than it's worth. For classical mc people probably implement their own models and lattices and tweak them as needed, rather than following whatever (lattice) interface we define. And for dqmc the lattice only matters for the setup, right?

carstenbauer replied:
> I wonder if this is more work than it's worth.

Maybe.

> For classical mc people probably implement their own models and lattices and tweak them as needed, rather than following whatever (lattice) interface we define.

The idea is that people only have to implement their model but could use/build upon a pre-defined lattice (from Lattices.jl, LatticePhysics.jl, or built-in). They don't really have to care about the `AbstractLattice` interface other than to see how they get the information from a lattice they need. In that sense, our `AbstractLattice` is basically just a synchronization of different `AbstractLattice` definitions and concrete lattice implementations out there. Sure, we could also just say, "if you use lattice library X look at their documentation for conventions and accessor functions". However, I think it's good to have a minimal common interface. This is also why I like functions like `neighbors_lattice_table(::AbstractLattice)`. After I call this function the lattice, independent of its origin, is just a lookup table that I know how to use.

> And for dqmc the lattice only matters for the setup, right?

Not quite, one also has to calculate the energy change in the bosonic sector for Metropolis. However, I agree that this is hardly a bottleneck in DQMC.

carstenbauer replied:
Of course, if Lattices.jl or LatticePhysics.jl were already mature/established, we wouldn't have to any of this. We could just support that single library and copy their conventions.

carstenbauer commented on src/models/HubbardAttractive/HubbardModelAttractive.jl:
```diff
@@ -82,20 +81,35 @@ Hubbard model in the DQMC simulation.
+
+    # Nearest neighbor hoppings
+    @inbounds @views begin
+        for (src, trg) in neighbors(l, Val(false))
```
See other comment: It would be better to construct a `neighbors_lookup_table` upfront.

crstnbr commented:
* [ ] LatPhys.jl: add `neighbor_lookup_table`

ffreyer commented on src/models/Ising/IsingModel.jl:
```diff
@@ -149,8 +154,10 @@ Calculate energy of Ising configuration `conf` for Ising model `m`.
     E = 0.0
-    @inbounds for (source, target) in neighbors(m.l)
-        E -= conf[source] * conf[target]
+    for n in 1:ndims(m)
```
This only works for cubic lattices. For example, triangular should have `n = 1:3`, honeycomb should alternate between `n = 1:1` and `n = 1:2`.

carstenbauer replied:
If you replace `ndims(m)` by `size(m.neighs, 1)` it should work for arbitrary lattices (that implement `neighbors_lookup_table(l)`).

ffreyer replied:
`size(m.neighs, 1)` would count every bond twice. It should be `div(size(m.neighs, 1), 2)`, but that is problematic if `size(m.neighs, 1)` is odd and it relies on `m.neighs` to be sorted in a specific way. (I.e. if `j in m.neighs[i, 1:div(end, 2)], then `i in m.neighs[i, div(end, 2):end]` must be true.)

carstenbauer replied:
You're right, my bad.

carstenbauer commented on src/models/Ising/IsingModel.jl:
```diff
@@ -153,6 +153,16 @@ end
 """
 function energy(mc::MC, m::IsingModel, conf::IsingConf)
+    E = 0.0
+    for (src, trg) in neighbors(m.l, Val(false))
```
You can drop the `Val(false)` here (it's the default) if you want.
